src/markdown_exec/formatters/jupyter.py

```diff
@@ -0,0 +1,253 @@
+                    if msg["parent_header"].get("msg_id") == msg_id:
+                        shutdown = True
+                        break
+                except queue.Empty:
+                    if not kc.is_alive():
+                        break
+                    if time.time() - start_time > _SHUTDOWN_TIMEOUT:
+                        shutdown = True
+                        break
+            kc.stop_channels()
+            km = _kernel_managers[language][key]
+            if not shutdown and km.is_alive():
+                # force shutdown if we didn't get a reply
+                km.shutdown_kernel()
+            km.cleanup_resources()
+    _kernel_clients.clear()
+    _kernel_managers.clear()
+
+# Because this is a language agnostic runner, it has two extra initial language
+# arguments. This makes it easy to use functools.partial to create a language
+# specific runner when needed.
+def _run_jupyter(
+    language: str,
+    start_kernel_kwargs: dict[str, Any] | None,
+    code: str,
+    returncode: int | None = None,
+    session: str | None = None,
+    id: str | None = None,  # noqa: A002, ARG001
+    **extra: str,
+) -> str:
+    """Execute code using a Jupyter kernel."""
+    is_named_session = session is not None and session != ""
+
+    # Get kernel manager and client
+    km, kc = _get_kernel_manager_and_client(language, session, start_kernel_kwargs)
+
+    kc.wait_for_ready()
+    # Send code for execution
+    msg_id = kc.execute(code, store_history=is_named_session)
+
+    # Collect output
+    outputs = []
+    error_outputs = []
+    execution_completed = False
+
+    while not execution_completed:
+        try:
+            # Check for shell messages (execution replies)
+            shell_msg = kc.get_shell_msg(timeout=0.1)
+            if shell_msg["parent_header"].get("msg_id") == msg_id:
+                execution_completed = True
+
+                if shell_msg["content"]["status"] == "error":
+                    if (
+                        shell_msg["content"]["traceback"] is not None
+                        and len(shell_msg["content"]["traceback"]) > 0
+                    ):
+                        if shell_msg["content"]["traceback"][0] == "Stack trace:":
+                            error_outputs.extend(
+                                shell_msg["content"]["traceback"][1:],
+                            )
+                        else:
+                            error_outputs.extend(shell_msg["content"]["traceback"])
+                    break
+                # note: abort is deprecated, but some older kernels may still send it
+                # it doesn't send error info, however - so we'll need to populate a fake abort error instead
+                if shell_msg["content"]["status"] == "abort":
+                    error_outputs.append("AbortError: Execution aborted")
+                    break
+        except queue.Empty:
+            pass
```

Comment:
lint: Same remark as before: what line exactly is susceptible of raising `queue.Empty`? Could we wrap only this line in `try`/`except`?